### PR TITLE
feature: container option

### LINKED ISSUE #63: Usage in non-laminas projects - custom container provider
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

I'd like to have a way to provide a container without having to create `config/container.php` or `config/application.config.php`.

When developing commands in a library, I'd like to have some kind of "integration" validation. 
Like: 
- is the command detected from the laminas CLI command
- how does my description look like when executing CLI command with `--help`
- e.g.

For now, I can create `config/container.php` or `config/application.config.php` and we're good to go. But I wonder if its possible to either add an option to a file where the container can be found or handle it like doctrine does with its `cli-config.php`.

I'd like to get some more ideas about this.

### PR DESCRIPTION
**Pull request by boesing**
|    Q          |   A
|-------------- | ------
| Documentation | yes
| BC Break      | no
| New Feature   | yes

### Description

With this feature, one can provide `--container=<path>` to provide a PSR-11 container which is not in a file called `config/container.php`.

The original idea behind this feature is, that I can create a generic PSR-11 container file which then parses the current project for a file `ConfigProvider` so it can automagically provide a container for that component without having to copy & paste that file to all components while creating a directory called `config` which has to be added to my global `.gitignore`.

```php
<?php
declare(strict_types=1);

if (!file_exists('composer.json')) {
    throw new RuntimeException('Missing `composer.json`!');
}

$composer = json_decode(file_get_contents('composer.json'), true, 512, JSON_THROW_ON_ERROR);
$configProviderClassName = $composer['extra']['laminas']['config-provider'] ?? '';
if ($configProviderClassName === '') {
    throw new RuntimeException('There is no laminas `config-provider` configuration within the `composer.json`!');
}

if (!class_exists($configProviderClassName)) {
    throw new RuntimeException('Laminas `config-provider` configuration within the `composer.json` contains invalid class name!');
}

$configProvider = new $configProviderClassName;
if (!is_callable($configProvider)) {
    throw new RuntimeException('Laminas `config-provider` is not callable!'); 
}

$config = $configProvider();
$dependencies = $config['dependencies'] ?? [];
$serviceManager = new \Laminas\ServiceManager\ServiceManager($dependencies);
$serviceManager->setService('config', $config);
$serviceManager->setAlias('Config', 'config');

return $serviceManager;
```

This actually works fine in almost all (laminas) components which provide a `config-provider` information within the `composer.json`. 
I can re-use that file stored wherever I like tho and have a configured service manager (until the components has it either required as production or dev dependency).

Imho, thats super helpful - may I should add this somewhere to the docs aswell?

Thoughts, @michalbundyra @weierophinney?

Fixes #63

### CODE REVIEW

**Comment by boesing** [comment]
This PR is based on #69 and thus contains changes from that PR.

File: README.md [hunk 1]
```diff
@@ -18,7 +18,7 @@ $ composer require laminas/laminas-cli
 ## Usage
 
 ```bash
-$ vendor/bin/laminas [command-name]
+$ vendor/bin/laminas [--container=<path>] [command-name]
```
Comment (michalbundyra) [inline README.md:20]: minor: imo option should be after argument:

```
vendor/bin/laminas [command-name] [--container=<path>]
```

File: composer.json [hunk 2]
```diff
@@ -38,6 +38,7 @@
         "laminas/laminas-coding-standard": "~2.0.0",
         "laminas/laminas-mvc": "^3.1.1",
         "laminas/laminas-servicemanager": "^3.4",
+        "mikey179/vfsstream": "2.0.x-dev",
```
Comment (michalbundyra) [inline composer.json:41]: any reason to use dev version? it might be unstable

File: docs/book/intro.md [hunk 3]
```diff
@@ -157,21 +156,21 @@ container to seed its application.
 To list all available commands, run:
 
 ```bash
-$ ./vendor/bin/laminas
+$ ./vendor/bin/laminas [--container=<path>]
```
Comment (michalbundyra) [inline docs/book/intro.md:161]: I am not sure if we should have it in all examples. I would drop it, as it makes things a bit more harder to read, imo.

File: src/ContainerResolver.php [hunk 4]
```diff
@@ -10,76 +10,108 @@
...
      *
      * @throws RuntimeException When cannot locate PSR-11 container for the application.
      */
-    public static function resolve(): ContainerInterface
-    {
-        if (file_exists('config/container.php')) {
-            return self::resolveDefaultContainer();
+    public function resolve(
+        InputInterface $input
+    ): ContainerInterface {
```
Comment (michalbundyra) [inline src/ContainerResolver.php]: no need to split into 3 lines

File: src/ContainerResolver.php [hunk 5]
```diff
@@ -10,76 +10,108 @@
...
+        $pathToContainer = $input->getOption(ApplicationFactory::CONTAINER_OPTION) ?? '';
+        assert(is_string($pathToContainer));
```
Comment (michalbundyra) [inline src/ContainerResolver.php]: I see in all other places we are using Assert:: not assert(, maybe also here we should be consistent then?

File: src/ContainerResolver.php [hunk 6]
```diff
@@ -99,4 +131,19 @@ private static function resolveMvcContainer(): ContainerInterface
...
         return $serviceManager;
     }
+
+    /**
+     * @psalm-param non-empty-string $containerPath
+     */
+    private function resolveContainerFromPath(string $containerPath): ContainerInterface
+    {
+        if (! file_exists($containerPath)) {
+            throw new InvalidArgumentException('Provided path must be relative to the project root.');
```
Comment (michalbundyra) [inline src/ContainerResolver.php]: I believe with absolute path it's going to work as well, we assume it's relative when absolute does not exist.

File: src/ContainerResolver.php [hunk 7]
```diff
@@ -99,4 +131,19 @@ private static function resolveMvcContainer(): ContainerInterface
...
+        }
+
+        $container = include $containerPath;
+        Assert::isInstanceOf($container, ContainerInterface::class, 'Failed to load PSR-11 container');
```
Comment (michalbundyra) [inline src/ContainerResolver.php:148]: maybe here, for the error message we should provide the path from where we try to load the container

File: test/ApplicationConfiguratorTest.php [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+ * @copyright https://github.com/laminas/laminas-cli/blob/master/COPYRIGHT.md
+ * @license   https://github.com/laminas/laminas-cli/blob/master/LICENSE.md New BSD License
+ */
+
+declare(strict_types=1);
+
+namespace LaminasTest\Cli;
+
+use Laminas\Cli\ApplicationConfigurator;
+use laminas\cli\containercommandloader;
```
Comment (michalbundyra) [inline test/ApplicationConfiguratorTest.php]: doesn't look right

**Review by michalbundyra — COMMENTED** [review]
@boesing  it's nice feature. I had a quick look on the PR and it's ok with some minor comments.
Just the main bin file is a far more complex now, but I believe it's needed.

I just wonder if it all works fine with chains? I wanted to avoid any global options but it makes sense to add this one.

Also the example from the PR description is nice, but in real world we would need to include also other providers from dependant libraries (for example for "satellite" packages etc). It would be good for using commands in the library (we had thought about it earlier, but it looks like nothing has been done in that direction).

File: composer.json (hunk 2 above)
Comment (boesing) [reply composer.json:41]: its only dev-dependency and thats the only version which currently supports PHP 8.0

File: src/ContainerResolver.php (hunk 5 above)
Comment (boesing) [reply src/ContainerResolver.php]: I use `assert`/`Assert` depending on what I want achieve.

--- 

When the result is always `true` and I only have to provide informations to psalm, I can either use `@var` annotation or I do an assertion which is mostly disabled in production systems.
The benefit in doing so is, that it has no runtime impact here and if something would change in the `InputInterface` implementation, it will directly fail in unit tests (which have `zend.assertion=1`).

---

In case the value I want to assert on is unknown to the LoC due to external dependency (e.g. integer which has to be positive but the annotation does not require that), I use `Assert::positiveInteger`.
So whenever a value is provided from upstream, I do not trust that and I enforce values with `Assert`.

---

In this case, we are defining the `container` option in the project. So there is most-likely **no** reason why to waste runtime assertion on that variable unless someone decides to somehow force the `InputInterface` implementation to not return string.

Comment (boesing) [reply src/ContainerResolver.php]: The best example here is something like this:

```php
$pathToContainer = sprintf('%s/%s', $this->projectRoot, $pathToContainer);
```

Psalm asserts that the `$pathToContainer` is just a string, even tho, it is absolutely not empty due to the fact that there is a `/` in there **and** that `projectRoot` property is non-empty-string. So there is **no** scenario where `$pathToContainer` would become an empty string.

So overall, when using `assert()`, that is just to provide proper details to psalm while using `Assert` would provide the same information to psalm but has runtime impact in upstream projects in production envs.

File: src/ContainerResolver.php (hunk 6 above)
Comment (boesing) [reply src/ContainerResolver.php]: True, it can be an absolute path aswell.
Actually, this method is re-used by the original implementation which just used `include`.
So in 2/3 use-cases, the variable is relative, but yes - with the `container` option I do also allow absolute paths.

Maybe I should use different `resolve` methods to handle that.

File: src/ContainerResolver.php (hunk 7 above)
Comment (boesing) [reply src/ContainerResolver.php:148]: Thats an existing message I've just copied.

I see what I can do.

File: test/ApplicationConfiguratorTest.php (hunk 8 above)
Comment (boesing) [reply test/ApplicationConfiguratorTest.php]: Good catch, dunno what happened there but I'll change that.

File: docs/book/intro.md (hunk 3 above)
Comment (boesing) [reply docs/book/intro.md:161]: Well, this would be once in the documentation and once in the README on github.
So actually, I think we should keep that.

File: README.md (hunk 1 above)
Comment (weierophinney) [reply README.md:20]: There's a concept of global options already in symfony/console, and those are generally injected before the command-name. (I've seen this in Composer, npm, and other systems as well, and this is how the option is already being implemented from what I can see.) I think specifying that it should be before the command name makes sense, as it configured the _laminas binary_, and not the command itself.

File: bin/laminas [hunk 9]
```diff
@@ -23,8 +27,33 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
...
+$definition = $app->getDefinition();
+$output = new ConsoleOutput();
+$containerNotFoundMessage = '';
+try {
+    $input = new ArgvInput();
+    $input->bind($definition);
+    $container = (new ContainerResolver($projectRoot))->resolve($input);
+    $app = (new ApplicationConfigurator())($app, $container);
+} catch (\Symfony\Component\Console\Exception\RuntimeException $exception) {
+    // Suppress as it will be part of the application output
```
Comment (weierophinney) [inline bin/laminas]: I'm confused here - wouldn't we want to report the issue with configuring the application and exit early for a configuration error like this? Or will symfony/console exit with an error status at this point anyways? Can you elaborate a bit on why it's okay to catch and do nothing here?

File: bin/laminas [hunk 10]
```diff
@@ -23,8 +27,33 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
...
+} catch (RuntimeException | InvalidArgumentException $exception) {
+    $containerNotFoundMessage = sprintf('<error>%s</error>', $exception->getMessage());
+} finally {
+    $input = new ArgvInput();
+}
+
+$exitCode = $app->run($input, $output);
```
Comment (weierophinney) [inline bin/laminas:54]: Why are we running the app if the container is not found?

File: src/ApplicationConfigurator.php [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+use Symfony\Component\Console\Application;
+use Symfony\Component\Console\ConsoleEvents;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Webmozart\Assert\Assert;
+
+/**
+ * @internal
+ */
+final class ApplicationConfigurator
```
Comment (weierophinney) [inline src/ApplicationConfigurator.php]: I really dislike this name - "Configurator" feels like a made-up Calvin and Hobbes term :rofl: .

Maybe `ApplicationProvisioner`?

File: src/ApplicationFactory.php [hunk 12]
```diff
@@ -30,33 +26,26 @@
...
-        // phpcs:ignore WebimpressCodingStandard.PHP.CorrectClassNameCase
-        $application->setCommandLoader(new ContainerCommandLoader($container, $commands));
-        $application->setDispatcher($dispatcher);
         $application->setAutoExit(false);
 
+        $definition = $application->getDefinition();
+        $definition->addOption(
+            new InputOption(
+                self::CONTAINER_OPTION,
+                'c',
```
Comment (weierophinney) [inline src/ApplicationFactory.php]: Let's not define a short-name for this. "c" is something I can see user commands wanting to use for an option (I can think of at least two commands I've written personally that use this short name off the top of my head).

File: src/ContainerResolver.php (hunk 5 above)
Comment (weierophinney) [reply src/ContainerResolver.php]: I think the runtime impact argument here is weak - this is the CLI, and the handful of assertions we'd be doing here are not likely to have an impact on the performance; the work the commands do themselves will be far more of an issue. I'd argue for consistency over performance in this context.

As an added bonus, webmozart/assert allows us to provide messages for failed assertions, which will then bubble up to the user.

File: test/ApplicationConfiguratorTest.php (hunk 8 above)
Comment (weierophinney) [reply test/ApplicationConfiguratorTest.php]: I've had psalm do this to me a few times when running `--alter`. Don't know why it does, and I've never been able to reliably reproduce the issue

**Review by weierophinney — CHANGES_REQUESTED** [review]
I like the idea behind this - it will definitely make it easier to test/run commands in individual packages as well by simply referencing a standard container file like you present in your patch description.

I've made a few comments, and noted one name I really dislike; change if you see fit. :smile:

File: bin/laminas (hunk 9 above)
Comment (boesing) [reply bin/laminas]: Uhm, okay I try.
When you call the "binary" with `--container`, some validation logic inside the `ArgvInput` of symfony throws that exception as the `--container` option is configured to require a value. 
This is, because we are actually accessing it in `ContainerResolver::resolve`.

As I dont want to duplicate that code which renders that exception to the console, I simply suppress the exception as it will be properly shown as well when executing `$app->run`.

So there are 3 options to me

1. catch exception, render the error as it would be done by the application by duplicating that code
2. catch exception, suppress it and let `$app->run` do its job
3. do not catch exception and have that default PHP "uncaught exception" error in CLI


Due to my laziness, I prefer the current solution (no. 2)

File: bin/laminas (hunk 10 above)
Comment (boesing) [reply bin/laminas:54]: The main idea behind this is, that you are able to see `--help` where all available options are listed (even tho, the new `--container` option).

Showing the user that no container was found is quite useless when you do not show a way how to fix that (e.g. by providing `--container`).

File: src/ApplicationConfigurator.php (hunk 11 above)
Comment (boesing) [reply src/ApplicationConfigurator.php]: I dont have a strong need for this name so sure, can rename that.

File: src/ContainerResolver.php (hunk 5 above)
Comment (boesing) [reply src/ContainerResolver.php]: Okay then, I use `assert` pretty much during my daily job to avoid useless runtime assertions but I can change that here.

--- 

So do we want to clarify how to handle this in future changes in other packages? I use `assert` a lot in cases where it should not be necessary to do any assertion (I gave the example with that `sprintf` LoC). If there is a coding guideline for laminas, we should add `assert` to forbidden functions in the codingstandard then?

File: README.md (hunk 1 above)
Comment (boesing) [reply README.md:20]: That was my intention as well. So I'll mark this as resolved as there is no TODO

File: test/ApplicationConfiguratorTest.php (hunk 8 above)
Comment (boesing) [reply test/ApplicationConfiguratorTest.php]: I did not use psalter here. So that was something else. Maybe an OSX issue with case insensitive filesystem...

File: bin/laminas (hunk 10 above)
Comment (boesing) [reply bin/laminas:54]: Thats why I provide the error message that the container could not be found afterwards.

File: test/ApplicationConfiguratorTest.php (hunk 8 above)
Comment (boesing) [reply test/ApplicationConfiguratorTest.php]: Okay I think thats the codestandard.
I probably used the FQCN within the code and the codestandard "fix" imported the use statement.
Since the `ContainerCommandLoader` does not really exist (due to the `class_alias` stuff which is done in the `autoload.php`), the codestandard seems to import it with lowercase characters.

File: bin/laminas (hunk 9 above)
Comment (weierophinney) [reply bin/laminas]: That's fine - just document what exactly you're suppressing here, so that future maintainers aren't left wondering why the code is the way it is. :)

File: bin/laminas (hunk 10 above)
Comment (weierophinney) [reply bin/laminas:54]: Makes sense. Again, I'd drop a note in here along the lines of "Running the app after suppressing container resolution exceptions allows symfony/console to report usage information".

File: src/ContainerResolver.php (hunk 5 above)
Comment (weierophinney) [reply src/ContainerResolver.php]: Something to bring up at the TSC meeting, I think, so we can then document it in the maintainers/contributor's guide(s).

File: autoload/autoload.php [hunk 13]
```diff
@@ -15,14 +15,13 @@
...
 
 /** @psalm-suppress DeprecatedClass */
 $version = strstr(Versions::getVersion('symfony/console'), '@', true) ?: '';
 
-if (str_replace('v', '', $version) >= '5.0.0') {
-    // phpcs:ignore WebimpressCodingStandard.PHP.CorrectClassNameCase
+if (version_compare(str_replace('v', '', $version), '5.0.0', '>=')) {
     class_alias(ContainerCommandLoaderTypeHint::class, ContainerCommandLoader::class);
 } else {
-    // phpcs:ignore WebimpressCodingStandard.PHP.CorrectClassNameCase
```
Comment (michalbundyra) [inline autoload/autoload.php]: I'd prefer to keep it here instead of excluding that rule globally. I am aware of the limitation when we using class_alias and class does not exist - then PHP thinks the class name is lowercase and so that sniff suggest it lowercased. I haven't found a way to detect that case and skip in the sniff.

File: composer.json [hunk 14]
```diff
@@ -38,12 +38,13 @@
...
         "phpunit/phpunit": "^9.4.1",
         "vimeo/psalm": "^4.4.1"
     },
     "autoload": {
         "files": [
-            "src/autoload.php"
+            "autoload/autoload.php"
```
Comment (michalbundyra) [inline composer.json]: minor: not sure why this change - if it's just consistency with other repos it's fine, but in general it should be separate PR...

File: src/ContainerResolver.php [hunk 15]
```diff
@@ -99,4 +129,43 @@ private static function resolveMvcContainer(): ContainerInterface
...
+     *
+     * @psalm-param non-empty-string $pathToContainer
+     */
+    private function isAbsolutePath(string $pathToContainer): bool
+    {
+        if ($pathToContainer[0] === '/') {
+            return true;
+        }
+
+        if (strpos($pathToContainer, '://') !== false) {
```
Comment (michalbundyra) [inline src/ContainerResolver.php:165]: is it working with remote container configuration?

File: test/ApplicationTest.php [hunk 16]
```diff
@@ -632,7 +625,7 @@ public function testHelpDisplaysInformationForCommandWithDependencies(): void
 
         $contains = [
             "Usage:\n  example:dep [options]\n",
-            "  -s, --string=STRING   A string option [default: \"default value\"]\n",
+            "  -s, --string=STRING        A string option [default: \"default value\"]\n",
```
Comment (michalbundyra) [inline test/ApplicationTest.php:628]: not sure why this change in test? Just different number of spaces?

**Review by michalbundyra — APPROVED** [review]
@boesing it's fine, just couple more minor comments, thanks

File: src/ContainerResolver.php (hunk 15 above)
Comment (boesing) [reply src/ContainerResolver.php:165]: Thats mainly for the unit tests where virtual streams are used, but yes, if the stream is supported by PHP, this would work. But actually I would not expose that as an "official" feature. Imho passing locally stored files should be preferred.

File: test/ApplicationTest.php (hunk 16 above)
Comment (boesing) [reply test/ApplicationTest.php:628]: Due to the new `--container` parameter, the indentation is not correct anymore and the test fails.

File: composer.json (hunk 14 above)
Comment (boesing) [reply composer.json]: Can move that to another PR if necessary. 
I personally just dont like autoload files being part of the "src" file as these do not follow any namespaces, e.g.

Imho, `src/` is meant to contain sources like interfaces, classes, traits, functions, e.g. So files containing `function` would be okay but files which do only provide autoloading polyfills should not be part of the `src/`.

I'm fine with keeping it as is aswell, just found it annoying.

File: autoload/autoload.php (hunk 13 above)
Comment (boesing) [reply autoload/autoload.php]: I personally would avoid having multiple ignore lines just because we think that this rule is somehow necessary in this project.
I can revert this and add dozens of `phpcs:ignore` but I am pretty sure that this will become a problem again somewhen in another PR.

Comment (Ocramius) [reply autoload/autoload.php]: The rule is valuable for long-term autoloading error prevention, whereas `class_alias()` stuff has always been the edge case.

Comment (boesing) [reply autoload/autoload.php]: Yes, but as we do reviews anyways, having that rule ignored in projects wont be a problem unless we fail reviewing features which implement new classes with invalid names. There are other rules which give more value than this one. 

But anyways, I'll revert these changes as I don't want to waste my time with discussions like this. I dont have a too strong opinion on this. 👍🏼 
Expect a PR later this day.

File: composer.json (hunk 14 above)
Comment (weierophinney) [reply composer.json]: Keeping it under `src/` keeps it consistent with our other repos. If you want to revisit that and suggest an alternative location that _all_ repos use, let's add that to the TSC agenda.

Comment (boesing) [reply composer.json]: @weierophinney which repositories do you mean?
https://github.com/laminas/laminas-cache/blob/2.12.x/autoload/patternPluginManagerPolyfill.php
https://github.com/laminas/laminas-form/blob/2.15.x/autoload/formElementManagerPolyfill.php

The only project I know besides of these is https://github.com/laminas/laminas-zendframework-bridge which has the `autoload.php` under `source`. 🤷🏼‍♂️ 

Anyways, I've reverted that change as I dont see any value in these discussions and neither have a too strong opinion on that. I am fine with "keep things as is", even tho they're not consistent.

**Comment by boesing** [comment]
@michalbundyra I've reverted the changes you noted. If the changes are fine for you and if @weierophinney gives me the okay, I'd add this to 1.1.0 and ship it.

File: bin/laminas [hunk 17]
```diff
@@ -23,8 +26,35 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
...
+$definition = $app->getDefinition();
+$output = new ConsoleOutput();
+$containerNotFoundMessage = '';
+try {
+    $input = new ArgvInput();
+    $input->bind($definition);
+    $container = (new ContainerResolver($projectRoot))->resolve($input);
+    $app = (new ApplicationProvisioner())($app, $container);
+} catch (\Symfony\Component\Console\Exception\RuntimeException $exception) {
+    // Running the app after suppressing container resolution exceptions allows symfony/console to report usage information
```
Comment (weierophinney) [inline bin/laminas:44]: :+1: 

Perfect!

File: bin/laminas [hunk 18]
```diff
@@ -23,8 +26,35 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
...
+} catch (RuntimeException | InvalidArgumentException $exception) {
+    // Usage information provided by the `ContainerResolver` should be passed to the CLI output
```
Comment (weierophinney) [inline bin/laminas:46]: :+1:

Definitely clarifies what is going on here.

File: bin/laminas [hunk 19]
```diff
@@ -23,8 +26,35 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
...
+    $containerNotFoundMessage = sprintf('<error>%s</error>', $exception->getMessage());
+} finally {
+    $input = new ArgvInput();
+}
+
+// By running the app even if its not provisioned allows symfony/console to report problems and/or display available options (like `--container`
```
Comment (weierophinney) [inline bin/laminas]:
```suggestion
// By running the app even if its not provisioned allows symfony/console to report problems
// and/or display available options (like `--container`)
```

File: docs/book/intro.md [hunk 20]
```diff
@@ -129,8 +129,7 @@ the default configuration and your [PSR-11 container](https://www.php-fig.org/ps
 cannot be detected automatically, you can still use the laminas-cli tool.
 
 To integrate such applications with laminas-cli, you will need to create a file
-that returns a PSR-11 container. Do so in the file `config/container.php`. Such
-a file might look like the following:
+that returns a PSR-11 container. Do so in the file `config/container.php` or if you already have such a file in another location, use the `--container=<path>` option. Such a file might look like the following:
```
Comment (weierophinney) [inline docs/book/intro.md]:
```suggestion
that returns a PSR-11 container.
Do so in the file `config/container.php` or if you already have such a file in another location, use the `--container=<path>` option.
Such a file might look like the following:
```

(new style guide suggests one sentence per line, so I've wrapped the new sentences to their own lines here)

**Review by weierophinney — CHANGES_REQUESTED** [review]
I can go ahead and merge these changes so we can release, if you'd like.

**Review by weierophinney — APPROVED** [review]
